Allow retries in Kafka Streams binder (#980)
spring-cloud/spring-cloud-stream-binder-kafka: docs/src/main/asciidoc/kafka-streams.adoc

However, you still can configure production exception handlers using the `StreamsBuilderFactoryBean` customizer which you can find more details about, in a subsequent section below.

=== Retrying critical business logic

There are scenarios in which you might want to retry parts of your business logic that are critical to the application.
There maybe an external call to a relational database or invoking a REST endpoint from the Kafka Streams processor.
These calls can fail for various reasons such as network issues or remote service unavailability.
More often, these failures may self resolve if you can try them again.
By default, Kafka Streams binder creates `RetryTemplate` beans for all the input bindings.

If the function has the following signature,
```
@Bean
public java.util.function.Consumer<KStream<Object, String>> process()
```
and with default binding name, the `RetryTemplate` will be registered as `process-in-0-RetryTemplate`.
This is following the convention of binding name (`process-in-0`) followed by the literal `-RetryTemplate`.

> In the case of multiple input bindings, there will be a separate `RetrtyTemplate` bean available per binding.

Reviewer: Typo still there `RetrtyTemplate`.

If there is a custom `RetryTemplate` bean available in the application and provided through `spring.cloud.stream.bindings.<binding-name>.consumer.retryTemplateName`, then that takes precedence over any input binding level retry template configuration properties.

> OOnce the `RetryTemplate` from the binding is injected into the application, it can be used to retry any critical sections of the application.

Reviewer: New typo `OOnce`.

Unlike the support for deserialization exception handlers as described above, the binder does not provide such first class mechanisms for handling production exceptions.
However, you still can configure production exception handlers using the `StreamsBuilderFactoryBean` customizer which you can find more details about, in a subsequent section below.

=== Retrying critical business logic

There are scenarios in which you might want to retry parts of your business logic that are critical to the application.
There maybe an external call to a relational database or invoking a REST endpoint from the Kafka Streams processor.
These calls can fail for various reasons such as network issues or remote service unavailability.
More often, these failures may self resolve if you can try them again.

> Kafka Streams binder provides an API that allows you to wrap the critical sections of the business logic in retrying temlate.

Reviewer: ```suggestion
The Kafka Streams binder provides an API that allows you to wrap the critical sections of the business logic in retrying template.
```

First, we show the usage of the API method that can only be used as a terminal operation.
This means that, once the method is called, your processing is complete and no further operations can be continued.
An example of a situation where this is useful is when you need to receive a stream of data, do some transformations and then insert the result into a relational database.
Here is a template to do that.

```
@Bean
public java.util.function.Consumer<KStream<Object, String>> process(CriticalPathRetryingKafkaStreamsDelegate<?, ?> criticalPathRetryingKafkaStreamsDelegate) {

    return input -> input
            .process(() -> new Processor<Object, String>() {
                @Override
                public void init(ProcessorContext processorContext) {
                }

                @Override
                public void process(Object o, String s) {
                    criticalPathRetryingKafkaStreamsDelegate.retryCriticalPath(() -> {
                        //critial business logic goes here.
                    });
                }

                @Override
                public void close() {
                }
            });
}
```

> As this is a termainl operation, it is recommended to use this API with low level Processor API in Kafka Streams.

Reviewer: ```suggestion
As this is a terminal operation, it is recommended to use this API with the low level Processor API in Kafka Streams.
```

> Here is another example of using a different API from this class, but this time it allows you to use it on non-termainal operation.

Reviewer: ```suggestion
Here is another example of using a different API from this class, but this time it allows you to use it on non-terminal operation.
```

However, you still can configure production exception handlers using the `StreamsBuilderFactoryBean` customizer which you can find more details about, in a subsequent section below.

=== Retrying critical business logic

There are scenarios in which you might want to retry parts of your business logic that are critical to the application.
There maybe an external call to a relational database or invoking a REST endpoint from the Kafka Streams processor.
These calls can fail for various reasons such as network issues or remote service unavailability.
More often, these failures may self resolve if you can try them again.
By default, Kafka Streams binder creates `RetryTemplate` beans for all the input bindings.

If the function has the following signature,
```
@Bean
public java.util.function.Consumer<KStream<Object, String>> process()
```
and with default binding name, the `RetryTemplate` will be registered as `process-in-0-RetryTemplate`.
This is following the convention of binding name (`process-in-0`) followed by the literal `-RetryTemplate`.

> In the case of multiple input bindings, there will be a separate `RetrtyTempate` bean available per binding.

Reviewer: Typo `RetrtyTempate`

> If there is a custom `RetryTemplate` bean is available in the application and provided through `spring.cloud.stream.bindings.<binding-name>.consumer.retryTemplateName`, then that takes precedence over any input binding level retry templates.

Reviewer: ```suggestion
If there is a custom `RetryTemplate` bean available in the application and provided through `spring.cloud.stream.bindings.<binding-name>.consumer.retryTemplateName`, then that takes precedence over any input binding level retry template configuration properties.
```

> Once the `RetryTemplate` from the binding is injected in the application, that can be used to retry any critical sections of the application.

Reviewer: ```suggestion
Once the `RetryTemplate` from the binding is injected into the application, it can be used to retry any critical sections of the application.
```

    @Bean
    @StreamRetryTemplate
    RetryTemplate fooRetryTemplate() {
        RetryTemplate retryTemplate = new RetryTemplate();

        RetryPolicy retryPolicy = new SimpleRetryPolicy(4);
        FixedBackOffPolicy backOffPolicy = new FixedBackOffPolicy();
        backOffPolicy.setBackOffPeriod(1);

        retryTemplate.setBackOffPolicy(backOffPolicy);
        retryTemplate.setRetryPolicy(retryPolicy);

        return retryTemplate;
    }

    @Bean
    public java.util.function.Consumer<KStream<Object, String>> process() {

        return input -> input
                .process(() -> new Processor<Object, String>() {
                    @Override
                    public void init(ProcessorContext processorContext) {
                    }

                    @Override
                    public void process(Object o, String s) {
                        fooRetryTemplate().execute(context -> {
                           //Critical business logic goes here.
                        });

                    }

                    @Override
                    public void close() {
                    }
                });
    }
}
```

> 

Reviewer: Note that when retries are exhausted, by default, the last exception will be thrown, causing the `KStream` to terminate.
If you wish to handle the exception and continue processing, you can add a `RecoveryCallback` to the execute method:

```
retryTemplate.execute(context -> {
                       //Critical business logic goes here.
                    }, context -> {
                       //Recovery logic goes here.
                       return null;
                    ));
```

Refer to the [spring-retry project](https://github.com/spring-projects/spring-retry) for more information about the `RetryTemplate`, retry policies, backoff policies and more.
